File: bungee/src/main/java/live/turna/phenyl/bungee/BungeePhenyl.java
package live.turna.phenyl.bungee;

import live.turna.phenyl.bungee.instance.*;
import live.turna.phenyl.bungee.listener.BungeeListenerManager;
import live.turna.phenyl.bungee.listener.MiraiListenerManager;
import live.turna.phenyl.common.config.Config;
import live.turna.phenyl.common.dependency.Log4jLoader;
import live.turna.phenyl.common.instance.PSender;
import live.turna.phenyl.common.listener.AbstractMiraiListenerManager;
import live.turna.phenyl.common.plugin.AbstractPhenyl;
import live.turna.phenyl.common.utils.MessageUtils;
import net.md_5.bungee.api.ProxyServer;
import net.md_5.bungee.api.plugin.Listener;
import net.md_5.bungee.api.plugin.Plugin;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicInteger;

public final class BungeePhenyl extends AbstractPhenyl {

    private final transient Plugin loader;
    private final transient MiraiListenerManager miraiListenerManager = new MiraiListenerManager();
    private final transient BungeeNativeLogger nativeLogger;
    private transient BungeeListenerManager bungeeListenerManager;
    private transient BungeeConfig bungeeConfig;
    private transient BungeeSenderFactory senderFactory;
    private transient BungeeForwarder forwarder;
    private transient BungeeMessenger messenger;

    public BungeePhenyl(Plugin loader) {
        this.loader = loader;
        nativeLogger = new BungeeNativeLogger(loader.getLogger());
    }

    public void onLoad() {
        super.load();
    }

    public void onEnable() {
        super.enable();
    }

    public void onDisable() {
        super.disable();
    }

    @Override
    public AbstractMiraiListenerManager getMiraiListenerManager() {
        return miraiListenerManager;
    }

    @Override
    protected boolean setupLog4j() {
        try {
            new Log4jLoader(this).onLoad();
            return true;
        } catch (IOException e) {
            e.printStackTrace();
            return false;
        }
    }

    @Override
    protected void initSenderFactory() {
        senderFactory = new BungeeSenderFactory(this);
    }

    @Override
    protected void initForwarder() {
        forwarder = new BungeeForwarder(this);
    }

    @Override
    protected void initMessenger() {
        messenger = new BungeeMessenger(this);
    }

    @Override
    protected void loadConfig() {
        bungeeConfig = new BungeeConfig(this);
        bungeeConfig.load();
    }

    @Override
    protected boolean postConfig() {
        return bungeeConfig.postLoad();
    }

    @Override
    protected void registerCommand() {
        ProxyServer.getInstance().getPluginManager().registerCommand(loader, new BungeeCommand(this, "phenyl", "ph"));
    }

    @Override
    protected void startListening() {
        bungeeListenerManager = new BungeeListenerManager(this);
        bungeeListenerManager.start();
    }

    @Override
    protected void stopListening() {
        bungeeListenerManager.end();
    }

    @Override
    protected void initMetrics() {
        new Metrics(loader, 14309);
    }

    @Override
    public BungeeForwarder getForwarder() {
        return forwarder;
    }

    @Override
    public BungeeMessenger getMessenger() {
        return messenger;
    }

    @Override
    public BungeeSenderFactory getSenderFactory() {
        return senderFactory;
    }

    @Override
    public String getVersion() {
        return loader.getDescription().getVersion();
    }

    @Override
    public Collection<PSender> getPlayers() {
        Collection<PSender> players = new ArrayList<>();
        ProxyServer.getInstance().getPlayers().forEach(player -> players.add(senderFactory.wrap(player)));
        return players;
    }

    @Override
    public String getPlatform() {
        return "BUNGEE";
    }

    @Override
    public File getDir() {
        return loader.getDataFolder();
    }

    @Override
    public BungeeNativeLogger getNativeLogger() {
        return nativeLogger;
    }

    @Override
    public PSender getPlayer(String username) {
        return this.getSenderFactory().wrap(ProxyServer.getInstance().getPlayer(username));
    }

    @Override
    public PSender getPlayer(UUID uuid) {
        return this.getSenderFactory().wrap(ProxyServer.getInstance().getPlayer(uuid));
    }

    @Override
    public HashMap<String, String> getOnlineList() {
        HashMap<String, String> result = new HashMap<>();
        ProxyServer.getInstance().getServers().forEach((s, serverInfo) -> {
            StringBuilder players = new StringBuilder();
            serverInfo.getPlayers().forEach(player -> players.append(player.getName()).append(","));
            if (players.isEmpty()) return;
            result.put(new MessageUtils(this).getServerName(serverInfo.getName()), players.substring(0, players.length() - 1));
        });
        return result;
    }

    @Override
    public CompletableFuture<HashMap<String, Boolean>> getStatus() {
        MessageUtils messageUtils = new MessageUtils(this);
        HashMap<String, Boolean> serverStatus = new HashMap<>();
        CompletableFuture<HashMap<String, Boolean>> statusFuture = new CompletableFuture<>();
        Config.enabled_servers.forEach(server -> {
            if (!ProxyServer.getInstance().getServers().containsKey(server)) {
                serverStatus.put(messageUtils.getServerName(server), false);
                return;
            }
            ProxyServer.getInstance().getServers().get(server).ping((result, error) -> {
                if (error == null) {
                    serverStatus.put(messageUtils.getServerName(server), true);
                } else
                    serverStatus.put(messageUtils.getServerName(server), false);
                if (serverStatus.size() == Config.enabled_servers.size())
                    statusFuture.complete(serverStatus);
            });
        });
        return statusFuture;
    }

    @Override
    public Integer getOnlineCount() {
        AtomicInteger onlineCount = new AtomicInteger();
        onlineCount.getAndSet(0);
        ProxyServer.getInstance().getServers().forEach((s, serverInfo) -> serverInfo.getPlayers().forEach(player -> onlineCount.getAndIncrement()));
        return onlineCount.get();
    }

    @Override
    public Boolean isProxy() {
        return true;
    }

    @Override
    public <T> void registerListener(T listener) {
        ProxyServer.getInstance().getPluginManager().registerListener(loader, (Listener) listener);
    }

    @Override
    public void unregisterListeners() {
        ProxyServer.getInstance().getPluginManager().unregisterListeners(loader);
    }

    @Override
    public InputStream getResourceAsStream(String name) {
        return loader.getResourceAsStream(name);
    }

    public Plugin getLoader() {
        return loader;
    }

}

File: bungee/src/main/java/live/turna/phenyl/bungee/BungeePhenyl.kt
package live.turna.phenyl.bungee

import live.turna.phenyl.bungee.instance.*
import live.turna.phenyl.bungee.listener.BungeeListenerManager
import live.turna.phenyl.bungee.listener.MiraiListenerManager
import live.turna.phenyl.common.config.Config
import live.turna.phenyl.common.dependency.Log4jLoader
import live.turna.phenyl.common.instance.PSender
import live.turna.phenyl.common.listener.AbstractMiraiListenerManager
import live.turna.phenyl.common.plugin.AbstractPhenyl
import live.turna.phenyl.common.utils.MessageUtils
import net.md_5.bungee.api.ProxyServer
import net.md_5.bungee.api.ServerPing
import net.md_5.bungee.api.config.ServerInfo
import net.md_5.bungee.api.connection.ProxiedPlayer
import net.md_5.bungee.api.plugin.Listener
import net.md_5.bungee.api.plugin.Plugin
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.util.*
import java.util.concurrent.atomic.AtomicInteger
import java.util.function.Consumer
import java.util.logging.Logger

class BungeePhenyl(val loader: Plugin) : AbstractPhenyl() {

    private val miraiListenerManager = MiraiListenerManager()

    private lateinit var bungeeListenerManager: BungeeListenerManager

    private lateinit var bungeeConfig: BungeeConfig

    private lateinit var senderFactory: BungeeSenderFactory

    private lateinit var forwarder: BungeeForwarder

    private lateinit var messenger: BungeeMessenger

    fun onLoad() {
        super.load()
    }

    fun onEnable() {
        super.enable()
    }

    fun onDisable() {
        super.disable()
    }

    override fun getMiraiListenerManager(): AbstractMiraiListenerManager {
        return miraiListenerManager
    }

    override fun setupLog4j(): Boolean {
        return try {
            Log4jLoader(this).onLoad()
            true
        } catch (e: IOException) {
            e.printStackTrace()
            false
        }
    }

    override fun initSenderFactory() {
        senderFactory = BungeeSenderFactory(this)
    }

    override fun initForwarder() {
        forwarder = BungeeForwarder(this)
    }

    override fun initMessenger() {
        messenger = BungeeMessenger(this)
    }

    override fun loadConfig() {
        bungeeConfig = BungeeConfig(this)
        bungeeConfig.load()
    }

    override fun postConfig(): Boolean {
        return bungeeConfig.postLoad()
    }

    override fun registerCommand() {
        ProxyServer.getInstance().pluginManager.registerCommand(loader, BungeeCommand(this, "phenyl", "ph"))
    }

    override fun startListening() {
        bungeeListenerManager = BungeeListenerManager(this)
        bungeeListenerManager.start()
    }

    override fun stopListening() {
        bungeeListenerManager.end()
    }

    override fun initMetrics() {
        Metrics(loader, 14309)
    }

    override fun getForwarder(): BungeeForwarder {
        return forwarder
    }

    override fun getMessenger(): BungeeMessenger {
        return messenger
    }

    override fun getSenderFactory(): BungeeSenderFactory {
        return senderFactory
    }

    override fun getVersion(): String {
        return loader.description.version
    }

    override fun getPlayers(): Collection<PSender> {
        val players: MutableCollection<PSender> = ArrayList()
        ProxyServer.getInstance().players.forEach(Consumer { player: ProxiedPlayer ->
            players.add(
                senderFactory.wrap(player)
            )
        })
        return players
    }

    override fun getPlatform(): String {
        return "BUNGEE"
    }

    override fun getDir(): File {
        return loader.dataFolder
    }

    override fun getNativeLogger(): Logger {
        return loader.logger
    }

    override fun getPlayer(username: String): PSender {
        return getSenderFactory().wrap(ProxyServer.getInstance().getPlayer(username))
    }

    override fun getPlayer(uuid: UUID): PSender {
        return getSenderFactory().wrap(ProxyServer.getInstance().getPlayer(uuid))
    }

    override fun getOnlineList(): HashMap<String, String> {
        val result = HashMap<String, String>()
        ProxyServer.getInstance().servers.forEach { (_: String?, serverInfo: ServerInfo) ->
            val players = StringBuilder()
            serverInfo.players.forEach(Consumer { player: ProxiedPlayer -> players.append(player.name).append(",") })
            if (players.isEmpty()) return@forEach
            result[MessageUtils(this).getServerName(serverInfo.name)] = players.substring(0, players.length - 1)
        }
        return result
    }

    override fun getStatus(): HashMap<String, Boolean> {
        val messageUtils = MessageUtils(this)
        val serverStatus = HashMap<String, Boolean>()
        Config.enabled_servers.forEach(Consumer { server: String? ->
            if (!ProxyServer.getInstance().servers.containsKey(server)) {
                serverStatus[messageUtils.getServerName(server)] = false
                return@Consumer
            }
            ProxyServer.getInstance().servers[server]!!
                .ping { _: ServerPing?, error: Throwable? ->
                    serverStatus[messageUtils.getServerName(server)] = error == null
                }
        })
        while (serverStatus.size != Config.enabled_servers.size) {
            try {
                Thread.sleep(100)
            } catch (e: InterruptedException) {
                if (Config.debug) e.printStackTrace()
            }
        }
        return serverStatus
    }

    override fun getOnlineCount(): Int {
        val onlineCount = AtomicInteger()
        onlineCount.getAndSet(0)
        ProxyServer.getInstance().servers.forEach { (_: String?, serverInfo: ServerInfo) ->
            serverInfo.players.forEach(
                Consumer { onlineCount.getAndIncrement() })
        }
        return onlineCount.get()
    }

    override fun isProxy(): Boolean {
        return true
    }

    override fun <T> registerListener(listener: T) {
        ProxyServer.getInstance().pluginManager.registerListener(loader, listener as Listener)
    }

    override fun unregisterListeners() {
        ProxyServer.getInstance().pluginManager.unregisterListeners(loader)
    }

    override fun getResourceAsStream(name: String): InputStream {
        return loader.getResourceAsStream(name)
    }
}